langchain/src/retrievers/multi_query.ts

```diff
@@ -0,0 +1,123 @@
+})
+
+// Export class
+export class MultiQueryRetriever extends BaseRetriever {
+    static lc_name() {
+        return "MultiQueryRetriever";
+    }
+    lc_namespace: string[] = ["langchain", "retrievers", "multiquery"];
+    private retriever: BaseRetriever;
+    private llmChain: LLMChain<LineList>;
+    private k:number;
+    private parserKey: string;
+
+    constructor({
+        retriever,
+        llmChain,
+        verbose = true,
+        k = 3, // The amount of different questions you'd like to generate. Provided to the prompt template
+        parserKey = "lines"
+    }: {
+        retriever: BaseRetriever
+        llmChain: LLMChain<LineList>
+        verbose?: boolean
+        k?: number
+        parserKey?: string
+    }) {
+        super();
+        this.retriever = retriever;
+        this.llmChain = llmChain;
+        this.verbose = verbose;
+        this.k = k;
+        this.parserKey = parserKey;
+    }
+
+    static fromLLM({
+        retriever,
+        llm,
+        prompt = DEFAULT_QUERY_PROMPT,
+        k = 3, // The amount of different questions you'd like to generate. Provided to the prompt template
+        parserKey = "lines"
+    }:{
+        retriever: BaseRetriever,
+        llm: BaseLLM,
+        prompt?: PromptTemplate,
+        k?:number, // The amount of different questions you'd like to generate. Provided to the prompt template
+        parserKey?:string
+    }): MultiQueryRetriever {
+        const outputParser = new LineListOutputParser();
+        const llmChain = new LLMChain({ llm, prompt, outputParser });
+        return new MultiQueryRetriever({ retriever, llmChain, k, parserKey });
+    }
+
+    private async generateQueries(question: string, k: number, runManager?: CallbackManagerForRetrieverRun): Promise<string[]> {
+        const response = await this.llmChain._call({ question, k }, runManager?.getChainChild());
+        const lines = response.text[this.parserKey] || [];
+        if (this.verbose) {
+            console.log(`Generated queries: ${lines}`);
+        }
+        return lines;
+    }
+
+    private async retrieveDocuments(queries: string[], runManager?: CallbackManagerForRetrieverRun): Promise<Document[]> {
+        const documents: Document[] = [];
+        for (const query of queries) {
+            const docs = await this.retriever.getRelevantDocuments(query, runManager?.getChild());
+            documents.push(...docs);
+        }
+        return documents;
+    }
+
+    private uniqueUnion(documents: Document[]): Document[] {
+        const uniqueDocumentsDict: { [key: string]: Document } = {};
+
+        for (const doc of documents) {
+            const key = `${doc.pageContent}:${JSON.stringify(Object.entries(doc.metadata).sort())}`;
+            uniqueDocumentsDict[key] = doc;
+        }
+
+        const uniqueDocuments = Object.values(uniqueDocumentsDict);
+        return uniqueDocuments;
+    }
+
+    // @ts-ignore
```

Comment:
What's this for?

Reply:
The different retrievals can (and most often will) return at least some identical documents. This function deduplicates them.

Reply:
No I mean the ts-ignore

Reply:
Ah sure, it wasn't necessary. Already removed it.